### PR TITLE
Rails::Auth::Env

### PR DESCRIPTION
**Pull request by tarcieri**
Factors all environment-related methods into a single class which can handle mutating the Rack environment.

All original getter/setter methods have been moved to `rails/auth/helpers.rb`

### CODE REVIEW

**Comment by tarcieri** [comment]
cc @ewr

File: lib/rails/auth/env.rb [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+      def authorize(allowed_by)
+        self.allowed_by = allowed_by
+        @authorized = true
+      end
+
+      # Set the name of the authority which authorized the request
+      #
+      # @param [String] :allowed_by label of what allowed the request
+      def allowed_by=(allowed_by)
+        raise AlreadyAuthorizedError, "already allowed by #{@allowed_by.ispect}" if @allowed_by
```
Comment (ewr) [inline lib/rails/auth/env.rb]: I suspect you meant `inspect` here.

Comment (tarcieri) [inline lib/rails/auth/env.rb]: Oops. Fixed in 06fdf9f

File: lib/rails/auth/credentials.rb [hunk 2]
```diff
@@ -1,45 +1,30 @@
...
-        credentials[type] = credential
+      def initialize(credentials = {})
+        raise TypeError, "expected Hash, got #{credentials.class}" unless credentials.is_a?(Hash)
+        @credentials = credentials
+      end
 
-        env
+      def []=(type, value)
+        raise TypeError, "expected String for type, got #{type.class}" unless type.is_a?(String)
+        raise AlreadyAuthorizedError, "credential '#{type}' has already been set" if @credentials.key?(type)
```
Comment (ewr) [inline lib/rails/auth/credentials.rb]: This removed the idempotent credential check. Was that intentional?

Comment (tarcieri) [inline lib/rails/auth/credentials.rb]: No, fixed in 3280815

**Pull request merged**